diff --git application/controllers/IcingadbimgController.php application/controllers/IcingadbimgController.php
index 90549e6..6c2cfe4 100644
--- application/controllers/IcingadbimgController.php
+++ application/controllers/IcingadbimgController.php
@@ -19,6 +19,7 @@ use Icinga\Module\Icingadb\Model\Service;
 use ipl\Stdlib\Filter;
 use ipl\Web\Url;
 
+#[\AllowDynamicProperties]
 class IcingadbimgController extends IcingadbGrafanaController
 {
     protected $host;
diff --git library/Grafana/ProvidedHook/Icingadb/HostDetailExtension.php library/Grafana/ProvidedHook/Icingadb/HostDetailExtension.php
index 85a0126..dc0ec75 100644
--- library/Grafana/ProvidedHook/Icingadb/HostDetailExtension.php
+++ library/Grafana/ProvidedHook/Icingadb/HostDetailExtension.php
@@ -8,6 +8,7 @@ use ipl\Html\Html;
 use ipl\Html\HtmlString;
 use ipl\Html\ValidHtml;
 
+#[\AllowDynamicProperties]
 class HostDetailExtension extends HostDetailExtensionHook
 {
     use IcingaDbGrapher;
diff --git library/Grafana/ProvidedHook/Icingadb/ServiceDetailExtension.php library/Grafana/ProvidedHook/Icingadb/ServiceDetailExtension.php
index 00fa6ed..0d4fa9c 100644
--- library/Grafana/ProvidedHook/Icingadb/ServiceDetailExtension.php
+++ library/Grafana/ProvidedHook/Icingadb/ServiceDetailExtension.php
@@ -8,6 +8,7 @@ use ipl\Html\Html;
 use ipl\Html\HtmlString;
 use ipl\Html\ValidHtml;
 
+#[\AllowDynamicProperties]
 class ServiceDetailExtension extends ServiceDetailExtensionHook
 {
     use IcingaDbGrapher;
